### PR TITLE
Update gradle to latest, add more flexibility

### PR DESCRIPTION
**Pull request by lipnitsk**
Update gradle wrapper to version 4.4, update the dependencies, use java-gradle-plugin to reduce boilerplate plugin code. Fix tests that were broken on Windows or due to old code.

Feature: Add flexibility by providing an ability to set custom command line options.

### CODE REVIEW

**Comment by lipnitsk** [comment]
@roberttaylor426 this is big one, but most churn is in build and test files, not actual plugin code. If you have idea on splitting things up I can definitely look into it. The plugin looks great, however I can't start using it in my tests without the extra functionality and flexibility that this request adds. Would be great to see these changes merged in and released when you have some time.
I hope my java-gradle-plugin changes and version updates did not break the publishing logic, but I wasn't sure how to test it.
You are also welcome to add the Travis CI badge into README.md, I didn't because the sourcemuse repo is not configured with Travis yet (should be only a couple of clicks to do it in Travis once you sign in with your GitHub credentials).

**Comment by roberttaylor426** [comment]
At first glance your proposed changes look pretty sensible, thanks for taking the time!

I'll have a closer look at each of the commits this evening and get back to you.

**Comment by dementiacs** [comment]
Can this branch be updated to pull in V2.0.0 of the flapdoodle stuff?

File: src/main/groovy/com/sourcemuse/gradle/plugin/GradleMongoPluginExtension.groovy [hunk 1]
```diff
@@ -25,6 +25,9 @@ class GradleMongoPluginExtension {
     String proxyHost = ''
     int proxyPort = 80
     String artifactStorePath = ''
+    boolean auth = false
+    Map<String, String> args = null
+    Map<String, String> params = null
```
Comment (roberttaylor426) [inline src/main/groovy/com/sourcemuse/gradle/plugin/GradleMongoPluginExtension.groovy]: Insofar as we have to explicitly check whether these properties are `null` at the call-site before acting, it might be preferable to default these to empty maps.

File: README.md [hunk 2]
```diff
@@ -47,13 +47,16 @@ The following properties are configurable:
...
 * ```port```: The port Mongo will listen on (defaults to **27017**). For random port assignment, set this value to **'RANDOM'** (the actual port value used will be available during the build through the **project.mongo.port** property)
 * ```proxyHost```: The proxy host name for Mongo downloads
 * ```proxyPort```: The proxy port for Mongo downloads
 * ```storageEngine```: The name of the storage engine to use. Can be **'wiredTiger'** or **'mmapv1'** for MongoDB Community Edition (default is **'wiredTiger'** for Mongo 3.2 and later; otherwise it is **'mmapv1'**). Alternative distributions might support additional engines.
 * ```storageLocation```: The directory location from where embedded Mongo will run, such as ```/tmp/storage``` (defaults to a java temp directory)
+* ```auth```: Enables [access control](https://docs.mongodb.com/manual/tutorial/enable-authentication/) (defaults to **false**)
+* ```params```: A <String,String> map of [MongoDB Server Parameters](https://docs.mongodb.com/manual/reference/parameters/) (defaults to **null**)
+* ```args```: A <String,String> map of mongod command-line options (value can be empty for arguments without values) (defaults to **null**)
```
Comment (roberttaylor426) [inline README.md]: Would you mind maintaining the alphabetical ordering of param documentation here?

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginTasksSpec.groovy [hunk 3]
```diff
@@ -232,33 +273,44 @@ class MongoPluginTasksSpec extends Specification {
...
-        gradleRunner.arguments << TEST_STOP_MONGO_DB
+        def args = TEST_STOP_MONGO_DB
 
         when:
-        gradleRunner.run()
+        GradleRunner.create()
+            .withPluginClasspath()
+            .withProjectDir(tmp.root)
+            .withArguments(args)
+            .build()
```
Comment (roberttaylor426) [inline src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginTasksSpec.groovy]: This chunk of code is duplicated 10 times in this test, and ultimately we only really care about the `args` in question. Could we extract a little helper method here, something like `runGradleWithArgs()`? It would probably then be fine to inline the `args` declaration.

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy [hunk 4]
```diff
@@ -282,13 +354,63 @@ class MongoPluginConfigSpec extends Specification {
...
+            .withPluginClasspath()
+            .withProjectDir(tmp.root)
+            .withArguments(args)
+            .build()
+        def mongoRunningDuringBuild = result.getOutput().contains(MONGO_RUNNING_FLAG)
+        def mongoRunningAfterBuild = mongoInstanceRunning()
+
+        then:
+        mongoRunningDuringBuild
+        !mongoRunningAfterBuild
```
Comment (roberttaylor426) [inline src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy]: Rather than checking that with authentication enabled there are no exceptions, could we test that it's actually having the desired effect? 

Could we attempt both authenticated and un-authenticated connections?

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy [hunk 5]
```diff
@@ -282,13 +354,63 @@ class MongoPluginConfigSpec extends Specification {
         new File(path).deleteDir()
     }
 
+    def 'authorization mode is enabled'() {
```
Comment (roberttaylor426) [inline src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy]: I think you meant 'authentication' rather than 'authorization'.

Also, could we rephrase this to read 'authentication mode can be enabled', so as not to mislead?

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy [hunk 6]
```diff
@@ -282,13 +354,63 @@ class MongoPluginConfigSpec extends Specification {
...
+        when:
+        GradleRunner.create()
+            .withPluginClasspath()
+            .withProjectDir(tmp.root)
+            .withArguments(args)
+            .build()
+
+        then:
+        noExceptionThrown()
+    }
```
Comment (roberttaylor426) [inline src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy]: I appreciate it might be tricky, but can we think of any `args` or `params` that might yield externally observable results that we could assert against, to verify the config is being correctly applied?

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy (hunk 4 above)
Comment (lipnitsk) [reply src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy]: Done

File: src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy (hunk 5 above)
Comment (lipnitsk) [reply src/test/groovy/com/sourcemuse/gradle/plugin/MongoPluginConfigSpec.groovy]: Reworded and added another test case

**Comment by lipnitsk** [comment]
Incorporated most of your feedback, still need to come up with observable results for `args` and `params` tests.

**Comment by roberttaylor426** [comment]
Great stuff. I'm going to merge this in and attempt to cut a release today.

Thanks again for your contribution.

**Comment by roberttaylor426** [comment]
I'm going to hold off on releasing until Flapdoodle release support for 3.6.0. [The code is now there](https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/pull/224), they just need to release a new version.

**Comment by lipnitsk** [comment]
Have you talked to Flapdoodle about when they will be making a release?

**Comment by roberttaylor426** [comment]
Initially I didn't as [the question had already been posed in a separate thread](https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/pull/224) and I didn't want to spam them, instead I upvoted.

As it's been a week I've now opened [a separate issue](https://github.com/flapdoodle-oss/de.flapdoodle.embed.mongo/issues/228). Let's see what they have to say.

**Comment by roberttaylor426** [comment]
1.0.4 is now out. Thanks again for your contribution.

**Comment by lperdigaoatTel** [comment]
Thank you guys for your good work. Very much appreciated :)
